fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs

alan-agius4 opened the pull request:
This commit implements a security fix to prevent XSS vulnerabilities where SVG animation elements (`<animate>`, `<set>`, etc.) could be used to modify the `href` or `xlink:href` attributes of other elements to `javascript:` URLs.

The fix introduces a runtime validation step:
- A new [ɵɵValidateAttribute](cci:1://file:///usr/local/google/home/alanagius/git/angular/packages/core/src/sanitization/sanitization.ts:276:0-288:1) instruction is used when `attributeName` is bound on SVG animation elements.
- If executed, a `RuntimeError` is thrown, preventing the binding.
- The compiler now identifies `attributeName` on SVG animation elements as security-sensitive and injects this validation.

Additionally, the DOM security schema has been updated to include a comprehensive list of MathML and SVG elements that accept `href` or `xlink:href` attributes, ensuring they are correctly treated as `SecurityContext.URL` and sanitized. This prevents malicious URLs from being bound to these attributes.


http://b/463880509

devversion reviewed (dismissed):
LGTM, but see comment above

alan-agius4 commented on packages/compiler/src/schema/dom_security_schema.ts:
```diff
@@ -32,21 +31,61 @@ export function SECURITY_SCHEMA(): {[k: string]: SecurityContext} {
+      'tref|href',
+      'tref|xlink:href',
+      'use|href',
+      'use|xlink:href',
```
1. I ask gemini to give me a comprehensive list of element where `href` can be used, after I validated the elements I added them. 
2. Yes I need to update the commit message, I restored not to use the `*` as it broke some tests.

devversion replied:
Potentially ISE maintains a similar list. I'm not sure how "complete" Gemini is here, but definitely an improvement.

cc. @bjarkler

alan-agius4 committed:
fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs

This commit implements a security fix to prevent XSS vulnerabilities where SVG animation elements (`<animate>`, `<set>`, etc.) could be used to modify the `href` or `xlink:href` attributes of other elements to `javascript:` URLs.

The fix introduces a runtime validation step:
- A new [ɵɵValidateAttribute](cci:1://file:///usr/local/google/home/alanagius/git/angular/packages/core/src/sanitization/sanitization.ts:276:0-288:1) instruction is used when `attributeName` is bound on SVG animation elements.
- If executed, a `RuntimeError` is thrown, preventing the binding.
- The compiler now identifies `attributeName` on SVG animation elements as security-sensitive and injects this validation.

Additionally, the DOM security schema has been updated to include a comprehensive list of MathML and SVG elements that accept `href` or `xlink:href` attributes, ensuring they are correctly treated as `SecurityContext.URL` and sanitized. This prevents malicious URLs from being bound to these attributes.

http://b/463880509

alan-agius4 committed:
fixup! fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs

alan-agius4 committed:
fixup! fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -273,3 +274,45 @@ function getSanitizer(): Sanitizer | null {
+ * @param tagName The name of the tag.
+ * @param attributeName The name of the attribute.
+ */
+export function ɵɵValidateAttribute(value: any, tagName: string, attributeName: string) {
```
Lowercasing instruction name for consistency:
```suggestion
export function ɵɵvalidateAttribute(value: any, tagName: string, attributeName: string) {
```

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -273,3 +274,45 @@ function getSanitizer(): Sanitizer | null {
+  const errorMessage =
+    ngDevMode &&
+    `Binding to attribute '${attributeName}' of '${tagName}' is disallowed for security reasons, ` +
+      `please use ${attributeName}="..."`;
```
```suggestion
      `please use regular static attributes instead: ${attributeName}="..."`;
```

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -273,3 +274,45 @@ function getSanitizer(): Sanitizer | null {
+
+/**
+ * @remarks Keep this in sync with DOM Security Schema.
+ * @see [SECURITY_SCHEMA](../../../compiler/src/schema/dom_security_schema.ts)
```
nit: we can add a similar note to the `dom_security_schema.ts` too.

AndrewKushnir reviewed (approved):
FYI, changing the status of the review from "Comment" to "Approve", since the feedback is not critical/blocking (see https://github.com/angular/angular/pull/65663#pullrequestreview-3520771793).

alan-agius4 committed:
fixup! fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs

alan-agius4 committed:
fixup! fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs

alan-agius4 committed:
fixup! fix(compiler): prevent XSS via SVG animation `attributeName` and MathML/SVG URLs
